[DOC] Update api stability documentation. (#2878)
seqan/seqan3: doc/about/api/index.md

There is no ABI stability.

Many of the rules on this page are derived from the [abseil library](https://abseil.io/about/compatibility).

# API stability {#api_stability}

In general, you can expect all **stable** entities within the `seqan3` namespace to be usable throughout the entire
release cycle of SeqAn3, i.e. if you write an application that includes `seqan-3.1.1`, you should be able
to compile it against `seqan-3.4.5` without errors.

Exceptions to the previous rule:
  -# **Do not depend upon internal details.** If something is in a namespace or filename/path that includes the word
     `detail` or if it is private to a class type, you are not allowed to depend upon it in any way. More generally,
     if it is not part of the *user documentation*, it is not part of the API.
  -# **Do not depend on entities marked as "experimental" or "no-api"**. Major changes to the library like new modules
     are often first marked as experimental within a minor release cycle. This means we do not guarantee
     stability until the next minor release happens and the experimental flag is removed in favour of the stable flag.

>      Some entities in namespace `seqan3` are permanently marked as "NOAPI" which designates them as subject to

Reviewer: ```suggestion
     Some entities in namespace `seqan3` are permanently marked as "no-api" which designates them as subject to
```

> to re-compile it against `seqan-3.4.5` without errors.

Author: I think you are only right about non-stable stuff. I think entities marked as stable that become deprecated at some point we promise to leave them in the code base until a new major release.

> to re-compile it against `seqan-3.4.5` without errors.

Reviewer: Yes, that's also right, stable stuff will stay until 4.0.0
But experimental entities might be deprecated and later removed?

> to re-compile it against `seqan-3.4.5` without errors.

Author: yes but that is already covered in the notes that follow ("do not depend on experimental stuff")

>      `detail` or if it is private to a class type, you are not allowed to depend upon it any way. More generally,

Reviewer: In line 1: `future-proofness`. Apparently, that's not really a word. There is `future-proof` and `future-proofing`, but not `future-proofness`.

Author: I came up with 

1. `# API Stability - future-proof promises of seqan3`
2. `# Future-proof API Stability`
3. `# Stability and long term promises`

Thoughts?

Reviewer: I like 3, though it should be `long-term`

Reviewer: > Ich finde 1. am besten. Begründung: 2. ist kaum unterscheidbar zu "# API stability" 3. weiß nicht, 1. gefällt mir besser :p

But `future-proof promises` means that the promises are future-proof, no?

Author: I kind of like 3 the most too. But I don't care much.

Reviewer 2: 2:1 for option three, I guess it is settled :-)

     if it is not part of the *user documentation*, it is not part of the API.
  2. **Do not depend on entities marked as "experimental" or "NOAPI"**. Major changes to the library like new modules

>      are often first marked as experimental within the a minor release cycle. This means we do not guarantee

Reviewer: ```suggestion
     are often first marked as experimental within a minor release cycle. This means we do not guarantee
```

     stability until the next minor release happens and the experimental flag is removed in favour of the stable flag.
     Some entities in namespace `seqan3` are permanently marked as "NOAPI" which designates them as subject to

>      unannounced change. In addition all entities in **core** and **utlity** are always marked "NOAPI". This is the

Reviewer: ```suggestion
     unannounced change. In addition, all entities in **core** and **utility** are always marked "NOAPI". This is the
```

     case for auxiliary data structures that would be part of `seqan3::detail` were they not needed to generate correct
     API documentation for entities in `seqan3` or considered nice-to-have general features for users.

>   4. **Do not depend on the *signatures* of SeqAn APIs.** In particular you may not take the address of APIs in SeqAn

Reviewer: ```suggestion
  4. **Do not depend on the *signatures* of SeqAn APIs.** In particular, you may not take the address of APIs in SeqAn
```

>      `detail` or if it is private to a class type, you are not allowed to depend upon it any way. More generally,

Reviewer 2: Nicht sicher, mein "Sprachgefühl" sagt mir, dort muss noch ein "in" oder ähnliches rein.
```suggestion
     `detail` or if it is private to a class type, you are not allowed to depend upon it in any way. More generally,
```

     if it is not part of the *user documentation*, it is not part of the API.
  2. **Do not depend on entities marked as "experimental" or "NOAPI"**. Major changes to the library like new modules
     are often first marked as experimental within a minor release cycle. This means we do not guarantee
     stability until the next minor release happens and the experimental flag is removed in favour of the stable flag.
     Some entities in namespace `seqan3` are permanently marked as "NOAPI" which designates them as subject to
     unannounced change. In addition, all entities in **core** and **utility** are always marked "NOAPI". This is the
     case for auxiliary data structures that would be part of `seqan3::detail` were they not needed to generate correct

>      API documentation for entities in `seqan3` or considered nice-to-have general features for users.

Author: agree but I changed it a little differently

>   4. **Do not depend on the *signatures* of SeqAn APIs.** In particular, you may not take the address of APIs in SeqAn

Author: changed everythin to automated numbering

  * The next (or another future) minor version removes the old API.

\warning

> As a special case of point 2. **the entire 3.0.* release is not stable**.

Reviewer: ```suggestion
As a special case of point 2, **the entire 3.0.* release is not stable**.
```

# Platform stability {#platform_stability}

The main requirement for being able to use SeqAn is a supported compiler and in general all operating systems that

> provide a supported compiler should be supported. In general, we only support the lates three major compiler versions.

Reviewer: ```suggestion
provide a supported compiler should be supported. In general, we only support the latest three major compiler versions.
```

> We currently support the following compilers on 64-bit operating systems with little-endian CPU architectures only:

Reviewer: ```suggestion
We currently support the following compilers on 64-bit operating systems with little-endian CPU architectures:
```

  * GCC7, GCC8, GCC9, GCC10, GCC11

\note Only the most recent minor release of a compiler is guaranteed to be supported, e.g. when `gcc-10.4` is released,
we may drop support for `gcc-10.3`. Since all platforms with an older version receive minor release updates,
this should not be a problem.

> We promise to support all of the above compilers until the last release of SeqAn3, or until all of the following

Reviewer: ```suggestion
We promise to support all the above compilers until the last release of SeqAn3, or until all the following
```

# Platform stability {#platform_stability}

The main requirement for being able to use SeqAn is a supported compiler and in general all operating systems that
provide a supported compiler should be supported. In general, we only support the latest three major compiler versions.

We currently support the following compilers on 64-bit operating systems with little-endian CPU architectures:
  * GCC7, GCC8, GCC9, GCC10, GCC11

\note Only the most recent minor release of a compiler is guaranteed to be supported, e.g. when `gcc-10.4` is released,
we may drop support for `gcc-10.3`. Since all platforms with an older version receive minor release updates,
this should not be a problem.

> We promise to support all the above compilers until the last release of SeqAn3, or until all the following

Reviewer 2: Nicht ganz sicher was damit ausgedrückt werden soll, vielleicht so?:
```suggestion
We promise to support the above compilers in the latest release of SeqAn3, or until all the following
```

Reviewer: I think it wants to express `Until SeqAn 4.0.0`. Though that's probably not correct anyway :)
